internal/experimental/registry/client.go

```diff
@@ -17,325 +17,476 @@ limitations under the License.
+// PullOptWithChart returns a function that sets the withChart setting on pull
+func PullOptWithChart(withChart bool) PullOption {
+	return func(operation *pullOperation) {
+		operation.withChart = withChart
 	}
-	c.printCacheRefSummary(r)
-	return r.Chart, nil
 }
 
-// RemoveChart deletes a locally saved chart
-func (c *Client) RemoveChart(ref *Reference) error {
-	r, err := c.cache.DeleteReference(ref)
-	if err != nil {
-		return err
+// PullOptWithProv returns a function that sets the withProv setting on pull
+func PullOptWithProv(withProv bool) PullOption {
+	return func(operation *pullOperation) {
+		operation.withProv = withProv
 	}
-	if !r.Exists {
-		return errors.New(fmt.Sprintf("Chart not found: %s", ref.FullName()))
+}
+
+// PullOptIgnoreMissingProv returns a function that sets the ignoreMissingProv setting on pull
+func PullOptIgnoreMissingProv(ignoreMissingProv bool) PullOption {
+	return func(operation *pullOperation) {
+		operation.ignoreMissingProv = ignoreMissingProv
 	}
-	fmt.Fprintf(c.out, "%s: removed\n", r.Tag)
-	return nil
 }
 
-// PrintChartTable prints a list of locally stored charts
-func (c *Client) PrintChartTable() error {
-	table := uitable.New()
-	table.MaxColWidth = c.columnWidth
-	table.AddRow("REF", "NAME", "VERSION", "DIGEST", "SIZE", "CREATED")
-	rows, err := c.getChartTableRows()
-	if err != nil {
-		return err
+type (
+	// PushOption allows specifying various settings on push
+	PushOption func(*pushOperation)
+
+	// PushResult is the result returned upon successful push.
+	PushResult struct {
+		Manifest *descriptorPushSummary         `json:"manifest"`
+		Config   *descriptorPushSummary         `json:"config"`
+		Chart    *descriptorPushSummaryWithMeta `json:"chart"`
+		Prov     *descriptorPushSummary         `json:"prov"`
+		Ref      string                         `json:"ref"`
 	}
-	for _, row := range rows {
-		table.AddRow(row...)
+
+	descriptorPushSummary struct {
+		Digest string `json:"digest"`
+		Size   int64  `json:"size"`
 	}
-	fmt.Fprintln(c.out, table.String())
-	return nil
-}
 
-// printCacheRefSummary prints out chart ref summary
-func (c *Client) printCacheRefSummary(r *CacheRefSummary) {
-	fmt.Fprintf(c.out, "ref:     %s\n", r.Name)
-	fmt.Fprintf(c.out, "digest:  %s\n", r.Manifest.Digest.Hex())
-	fmt.Fprintf(c.out, "size:    %s\n", byteCountBinary(r.Size))
-	fmt.Fprintf(c.out, "name:    %s\n", r.Chart.Metadata.Name)
-	fmt.Fprintf(c.out, "version: %s\n", r.Chart.Metadata.Version)
-}
+	descriptorPushSummaryWithMeta struct {
+		descriptorPushSummary
+		Meta *chart.Metadata `json:"meta"`
+	}
 
-// getChartTableRows returns rows in uitable-friendly format
-func (c *Client) getChartTableRows() ([][]interface{}, error) {
-	rr, err := c.cache.ListReferences()
+	pushOperation struct {
+		provData   []byte
+		strictMode bool
+	}
+)
+
+// Push uploads a chart to a registry.
+func (c *Client) Push(data []byte, ref string, options ...PushOption) (*PushResult, error) {
```

Comment:
It looks like PushResult is only used in tests.  Should this information be displayed to the user?  If not it should probably be removed from the public API.

Reply:
Maybe when using `--debug`?

Reply:
I'm considering the usage of this package outside the scope of Helm. Users may want to operate on the digests etc. when creating their own clients.

Due to the way the Pusher is implemented (based on Getter), there is no easy way to write to terminal from the `Push` method. So instead, I'm writing to `c.out`, which some people may set to /dev/null when using their own client